### PR TITLE
replaced listObjects with listObjectsV2 to overcome the 1K file sync …

### PR DESCRIPTION
**Pull request by rmoskal**
Ok, I replaced listObjects with listObjectsV2 and return the entire results set in a stream.

I tried to follow your code style. It would have been nice to use some E6 features (fat arrow functions and some destructuring). I tried to be minimally disruptive!  converter is not necessary any more.

I made the function I added  a tiny bit fancier than I needed to (you can pass in a mapping function). I did not test  the case where you don't pass in a function. I started writing some, but had some trouble with tests leaving stuff in the S3 bucket.  If you are uncomfortable with this, I'll replace it with a version that returns only the key.

I did this because I need it!  I'm surprised other people aren't barking more loudly!

Let me know what else you need!

### CODE REVIEW

**Comment by pioug** [comment]
I think we need your `s3-pagable-stream.js` file 😆

**Comment by rmoskal** [comment]
Guess I was trying to keep my commit log too clean!

**Comment by pioug** [comment]
Your changes looks good! I added the other maintainers as reviewers. I'll let some time for them to check just in case but eventually I'll probably merge your PR. Thanks @rmoskal 💪

File: lib/s3-pagable-stream.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * like S3::listObjectsV2()
+ *
+ * @param {function} req - a non executed AWS function
+ * @param {function} fn - a function that selects/maps the results
+ * @param {object} opts - stream options
+ */
+module.exports = function(req, fn, opts) {
+  opts = Object.assign({}, opts, { read: read, objectMode: true });
+  if (!fn)
+    fn = function(_in) {
```
Comment (pgherveou) [inline lib/s3-pagable-stream.js]: I would name it `identity` and move it at the top of the module

File: lib/s3-pagable-stream.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      return _in;
+    };
+
+  var stream = new Readable(opts);
+  stream.on('error', function(e) {
```
Comment (pgherveou) [inline lib/s3-pagable-stream.js]: I think we can remove this and let the consumer the responsibility of handling error

File: lib/converter.js [hunk 3]
```diff
@@ -1,13 +0,0 @@
-var xmlNodes = require('xml-nodes');
-var xmlObjects = require('xml-objects');
-var pumpify = require('pumpify');
```
Comment (pgherveou) [inline lib/converter.js]: these packages are used elsewhere or can be removed from the package.json?

File: lib/index.js [hunk 4]
```diff
@@ -1,5 +1,5 @@
 var AWS = require('aws-sdk'),
-  converter = require('./converter'),
+  s3pagableStream = require('./s3-pagable-stream'),
```
Comment (pioug) [inline lib/index.js]: It should be spelled "pageable", right? 🤔

Comment (rmoskal) [reply lib/index.js]: Let's call it paging-stream, since that's more like what it does.

File: lib/s3-pagable-stream.js (hunk 2 above)
Comment (rmoskal) [reply lib/s3-pagable-stream.js]: sounds good :).

File: lib/converter.js (hunk 3 above)
Comment (rmoskal) [reply lib/converter.js]: I will check.

File: lib/s3-pagable-stream.js (hunk 1 above)
Comment (rmoskal) [reply lib/s3-pagable-stream.js]: I'm not following what you want here. You want that to be a named function called identity?

Comment (rmoskal) [reply lib/s3-pagable-stream.js]: Oh, I see. I thought about doing that.

**Pull request merged**